src/compiler/checker.ts

```diff
@@ -22767,7 +22767,15 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return propType;
         }
         if (everyType(type, isTupleType)) {
-            return mapType(type, t => getRestTypeOfTupleType(t as TupleTypeReference) || undefinedType);
+            return mapType(type, t => {
+                const tupleType = t as TupleTypeReference;
+                const restType = getRestTypeOfTupleType(tupleType);
+                if (!restType) {
+                    return undefinedType;
+                }
+                const possibleOutOfBounds = index >= tupleType.target.fixedLength + getEndElementCount(tupleType.target, ElementFlags.Fixed);
+                return compilerOptions.noUncheckedIndexedAccess && possibleOutOfBounds ? getUnionType([restType, undefinedType]) : restType;

```

Comment:
`getEndElementCount` iterates over each element of a tuple. I'd rather check noUncheckedIndexAccess first:

```suggestion
                if (compilerOptions.noUncheckedIndexedAccess &&                
                    index >= tupleType.target.fixedLength + getEndElementCount(tupleType.target, ElementFlags.Fixed) {                    
                    return getUnionType([restType, undefinedType]);                    
                }                
                else {                
                    return restType;                    
                }
```

Losing the name for `possibleOutOfBounds` isn't that bad; I think just about every programmer will recognise `index => ...` as checking for that.